### PR TITLE
Fixes for Windows

### PR DESCRIPTION
**Pull request by dmikushin**

### CODE REVIEW

File: src/bench2.c [hunk 1]
```diff
@@ -19,7 +19,7 @@
 #define BENCH_OUTLEN 16
 #define BENCH_INLEN 16
 
-static double min(const double *samples, size_t count)
+static double min_(const double *samples, size_t count)
```
Comment (WOnder93) [inline src/bench2.c]: Could you please rename the function to `pick_min` instead? Other than that I'm fine with the changes. Thanks!

**Comment by dmikushin** [comment]
Hi @WOnder93, pick_min it is. I've also added another change with annotates exported function on Windows. Unlike on Linux, on Windows dll by default exports nothing.

File: include/argon2.h [hunk 2]
```diff
@@ -13,6 +13,16 @@
...
 
+#ifdef WIN32
+#ifdef argon2_EXPORTS
+#define ARGON2_API __declspec(dllexport)
+#else
+#define ARGON2_API __declspec(dllimport)
+#endif
+#else
+#define ARGON2_API
+#endif // WIN32
```
Comment (WOnder93) [inline include/argon2.h]: There is already some dllexport mechanism defined a few lines below - perhaps you could reuse/improve that?

File: lib/core.h [hunk 3]
```diff
@@ -134,7 +134,7 @@ void secure_wipe_memory(void *v, size_t n);
  * @param mem Pointer to the memory
  * @param s Memory size in bytes
  */
-void clear_internal_memory(void *v, size_t n);
+ARGON2_API void clear_internal_memory(void *v, size_t n);
```
Comment (WOnder93) [inline lib/core.h]: Why do you want to export `clear_internal_memory`? AFAIK it should be only used by the library code itself.

File: include/argon2.h (hunk 2 above)
Comment (WOnder93) [reply include/argon2.h]: (FYI, it is inherited from the original Argon2 code, I am not familiar with the peculiarities of building libraries on Windows...)

File: lib/core.h (hunk 3 above)
Comment (dmikushin) [reply lib/core.h]: From my perspective it looks like this. There are 3 projects: `argon2`, `argon2-exec` and `argon2-test`. The first one defines `clear_internal_memory`, two others make use of it in `test.c` and `run.c`.  Hence, there is a link error without export.

Comment (WOnder93) [reply lib/core.h]: Oh, right, those two use it, too... The encapsulation in the original code is just not very good :/ I guess for now it is best to keep it simple and just mark `clear_internal_memory` with the exisitng `ARGON2_PUBLIC` (if that is enough).

**Comment by dmikushin** [comment]
@WOnder93, if everything is OK, could you please kindly merge this request? Thanks!

**Comment by WOnder93** [comment]
@dmikushin, please address the comment on `include/argon2.h` first - I think you should reuse the existing `ARGON2_PUBLIC` macro (and optionally add there the dllimport case if you believe that it's necessary).

Thanks!

**Comment by dmikushin** [comment]
Oh thanks, I've missed that. Now it's done, could you please take a look?

A nice thing about CMake is that the library target automatically declares `${TARGET_NAME}_EXPORTS` macro, which could be used to determine the direction of dllimport/dllexport.

**Comment by WOnder93** [comment]
That looks good, thanks!

I knew about the CMake exports macro, but I was trying to keep the code independent of the build system and I also didn't want to diverge from the original code too much (should I decide to backport the changes there).

Anyway, I committed it as-is and added an explicit define of `argon2_EXPORT` into the QMake project. Maybe I'll add it to the Autotools macros later, but that is probably not going to be used on Windows anyway...

Merged manually with 57478dff430e8906560f75db57b24bc3a2892615..14e43c986fcba8c88f4c26c71980779f3aae68ae

**Comment by dmikushin** [comment]
Awesome, thanks!

**Pull request closed without merging**